cmd/kindnetd/main.go

```diff
@@ -100,43 +151,69 @@ func main() {
 }
 
 // nodeNodesReconciler returns a reconciliation func for nodes
-func makeNodesReconciler(cniConfig *CNIConfigWriter, hostIP string) func(*corev1.NodeList) error {
+func makeNodesReconciler(cniConfig *CNIConfigWriter, hostIP string, ipFamily IPFamily, clientset *kubernetes.Clientset) func(*corev1.NodeList) error {
 	// reconciles a node
 	reconcileNode := func(node corev1.Node) error {
-		// first get this node's IP
-		nodeIP := internalIP(node)
-		if nodeIP == "" {
-			fmt.Printf("Node %v has no Internal IP, ignoring\n", node.Name)
-			return nil
+		// first get this node's IPs
+		// we don't support more than one IP address per IP family for simplification
+		nodeIPs := internalIPs(node)
+		var nodeIPv4, nodeIPv6 string
+		for _, ip := range nodeIPs.List() {
+			if utilsnet.IsIPv6String(ip) {
+				nodeIPv6 = ip
+			} else {
+				nodeIPv4 = ip
+			}
 		}
 
 		// don't do anything unless there is a PodCIDR
-		podCIDR := node.Spec.PodCIDR
-		if podCIDR == "" {
+		var podCIDRs []string
+		if ipFamily == DualStackFamily {
+			podCIDRs = node.Spec.PodCIDRs
+		} else {
+			podCIDRs = []string{node.Spec.PodCIDR}
+		}
+		if len(podCIDRs) == 0 {
 			fmt.Printf("Node %v has no CIDR, ignoring\n", node.Name)
 			return nil
 		}
-
+		podCIDRsv4, podCIDRsv6 := splitCIDRs(podCIDRs)
 		// This is our node. We don't need to add routes, but we might need to
 		// update the cni config.
-		if nodeIP == hostIP {
+		if nodeIPs.Has(hostIP) {
 			fmt.Printf("handling current node\n")
 			// compute the current cni config inputs
 			if err := cniConfig.Write(
 				ComputeCNIConfigInputs(node),
 			); err != nil {
 				return err
 			}
+			// update internal IPs node annotations until #42125 is fixed
+			// TODO: https://github.com/kubernetes/kubernetes/issues/4A2125
+			// in kind we can hardcode the outer interface but we can be smarter
+			// and find the interface matching the HOST_IP
+			node.Annotations["io.k8s.sigs.kind.kindnet/addresses"] = getNodeExternalIPs(hostIP)
```

Comment:
this is kinda gross :/

Reply:
I know, but ... what are the alternatives to https://github.com/kubernetes/kubernetes/issues/42125?
using kubelet with the controller=external?

Reply:
yeah I think so, I'm exploring that now.

Reply:
Using CRDs?